contracts/protocol/v1/proxies/P1SoloBridgeProxy.sol

```diff
@@ -0,0 +1,454 @@
+        "address perpetual,",
+        "uint256 soloAccountNumber,",
+        "uint256 soloMarketId,",
+        "bool toPerpetual,",
+        "uint256 amount,",
+        "uint256 expiration,",
+        "bytes32 salt",
+        ")"
+    ));
+
+    // ============ Structs ============
+
+    struct Transfer {
+        address account;
+        address perpetual;
+        uint256 soloAccountNumber;
+        uint256 soloMarketId;
+        bool toPerpetual; // Indicates whether the transfer is from Solo to Perpetual or vice versa.
+        uint256 amount;
+        uint256 expiration;
+        bytes32 salt;
+    }
+
+    // ============ Events ============
+
+    event LogTransferred(
+        address indexed account,
+        address perpetual,
+        uint256 soloAccountNumber,
+        uint256 soloMarketId,
+        bool toPerpetual,
+        uint256 amount
+    );
+
+    event LogTransferCanceled(
+        address indexed account,
+        bytes32 transferHash
+    );
+
+    // ============ Immutable Storage ============
+
+    // Address of the Solo margin contract.
+    address public _SOLO_MARGIN_;
+
+    // Hash of the EIP712 Domain Separator data
+    bytes32 public _EIP712_DOMAIN_HASH_;
+
+    // ============ Mutable Storage ============
+
+    // transfer hash => bool
+    mapping (bytes32 => bool) public _HASH_USED_;
+
+    // ============ Constructor ============
+
+    constructor (
+        address soloMargin,
+        uint256 chainId
+    )
+        public
+    {
+        _SOLO_MARGIN_ = soloMargin;
+
+        /* solium-disable-next-line indentation */
+        _EIP712_DOMAIN_HASH_ = keccak256(abi.encode(
+            EIP712_DOMAIN_SEPARATOR_SCHEMA_HASH,
+            keccak256(bytes(EIP712_DOMAIN_NAME)),
+            keccak256(bytes(EIP712_DOMAIN_VERSION)),
+            chainId,
+            address(this)
+        ));
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Sets the maximum allowance on the Solo contract for a given market. Must be called
+     *  at least once on a given market before deposits can be made.
+     * @dev Cannot be run in the constructor due to technical restrictions in Solidity.
+     */
+    function approveMaximumOnSolo(
+        uint256 soloMarketId
+    )
+        external
+    {
+        address solo = _SOLO_MARGIN_;
+        IERC20 tokenContract = IERC20(I_Solo(solo).getMarketTokenAddress(soloMarketId));
+
+        // safeApprove requires unsetting the allowance first.
+        tokenContract.safeApprove(solo, 0);
+
+        // Set the allowance to the highest possible value.
+        tokenContract.safeApprove(solo, uint256(-1));
+    }
+
+    /**
+     * @notice Sets the maximum allowance on the Perpetual contract. Must be called at least once
+     *  on a given Perpetual before deposits can be made.
+     * @dev Cannot be run in the constructor due to technical restrictions in Solidity.
+     */
+    function approveMaximumOnPerpetual(
```

Comment:
Nit: Maybe we can abstract this out into a `P1Proxy` base contract. I think I will need this function in the future for other proxy contracts as well (e.g. for wrapping ETH into WETH, and then depositing)